### ISSUE TITLE
Dialog focus not restoring to a menu trigger based on react-aria example

### ISSUE THREAD

**Issue by snowystinger** [adobe/react-spectrum#3601]
# 🐛 Bug Report 

Discussion started here https://github.com/adobe/react-spectrum/pull/3323#issuecomment-1257231881

A Dialog inside a Menu is not restoring focus to the Menu's trigger when the Dialog closes.

## 🔦 Context

Don't know exactly why yet, but in the example, FocusScope's removeTreeNode for the Menu runs before the useLayoutEffect which assigns a nodeToRestore in useRestoreFocus.
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/focus/src/FocusScope.tsx#L789 is undefined, and I don't think it'd matter if we checked if this was a restoring scope instead of nodeToRestore existing, it'd be overwritten in the useRestoreFocus layoutEffect to the wrong value.

In our component, it's reversed. There seems to be a substantial delay between closing of the menu and actually unmounting it. However, I don't quite know where that delay is coming from, except that we use Transition instead of CSSTransition. I don't know if that is what makes the difference.

I pushed a branch with the sandbox code added to one of our stories for easier debugging https://github.com/adobe/react-spectrum/compare/focusscope-nodetorestore-wrong?expand=1

Reporter:
> I tried the new release and the issue seems to persist, I updated the sandbox. Also I removed CSSTransition to rule it out. I wonder if I constructed the test case wrong, I copied the files from the two official styled examples but some imports were changed so there may potentially be some API changes that didn't make it to the sandbox. The reason I doubt it is because the release announcement example on Twitter seems to be working nicely.

## 💻 Code Sample

https://codesandbox.io/s/sad-platform-58uy0k?file=/src/App.tsx

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |
| Browser          |
| Operating System |

**Comment by majornista** [comment]
@frontsideair, I have been able to provide a workaround for the issue you describe with a few changes to the example you provide at https://codesandbox.io/s/sad-platform-58uy0k?file=/src/App.tsx. 

1. I wrap the Popover used for the Menu in a CSSTransition with `in={isOpen}`: https://codesandbox.io/s/exciting-hawking-dsnlzr?file=/src/Popover.tsx:885-1500. This way the CSSTransition will take care or hiding and showing the Popover containing the Menu. Note, I also moved the FocusScope to be contained within the `popoverRef` `div`, so that the classNames used for animating the opacity get added to the Popover and not the hidden FocusScope sentinel.

```js
  return (
    <CSSTransition
      in={isOpen}
      mountOnEnter
      unmountOnExit
      timeout={{ enter: 0, exit: 0 }}
      classNames={{
        enter: "opacity-0",
        enterDone: "opacity-1 transition ease-in",
        exit: "opacity-0 transition ease-out"
      }}
    >
      <div
        {...overlayProps}
        ref={popoverRef}
        className="absolute z-10 top-full min-w-full shadow-lg border border-gray-300 bg-white rounded-md mt-2"
      >
        <FocusScope restoreFocus>
          {children}
          <DismissButton onDismiss={onClose} />
        </FocusScope>
      </div>
    </CSSTransition>
  );
```

2. I remove the brute force `state.isOpen` boolean at https://codesandbox.io/s/sad-platform-58uy0k?file=/src/Menu.tsx:1227-1500 and replace it by passing `state.isOpen` as the `isOpen` prop on the Popover at https://codesandbox.io/s/exciting-hawking-dsnlzr?file=/src/Menu.tsx:1299-1537.

This: 
```js
      {state.isOpen && (
        <Popover isOpen onClose={state.close}>
          <Menu
            {...menuProps}
            {...props}
            autoFocus={state.focusStrategy || true}
            onClose={() => state.close()}
          />
        </Popover>
      )}
```

Becomes this:
```js
      <Popover isOpen={state.isOpen} onClose={state.close}>
        <Menu
          {...menuProps}
          {...props}
          autoFocus={state.focusStrategy || true}
          onClose={() => state.close()}
        />
      </Popover>
```

I think this bug is still valid, but hopefully these small changes will unblock you.

**Comment by frontsideair** [comment]
This seems to be working in the sandbox, thanks. I only had to change `usePreventScroll({ isDisabled: !props.isOpen })`, there wasn't enough content in the page to show it's not scrollable.

In my app though it didn't solve the problem. It could be because I used a ternary like `return props.isOpen ? <div... : null` instead of `CSSTransition` but I believe they should behave similarly due to it unmounting on exit with zero timeout but I can't be too sure. I'll tweak more. 

My current workaround is to use a custom `useModalState` that takes the menu button ref and wrap `onClose` to add a timed out focus.
